### PR TITLE
Allow supplying custom "Expression" values to `@test`

### PR DESCRIPTION
**Pull request by exaexa**
This looks extremely helpful to me when testing a lot of "generated" test cases, and does not seem to add much breakable code.

use-case:
```julia
for something in somethings
  x = tricky_f(something)
  @test looks_ok(x, something)
end
```

All reports in above code give a relatively boring outcome of "Expression `looks_ok(x, something)` is somehow broken."

With this patch, users may append the "explanation" as

```julia
  [...]
  @test looks_ok(x,something)  expr=:(looks_ok($x,$something))
  [...]
```
which gives much more desirable values, and the test report directly says e.g. "Expression `looks_ok(1,2)` is broken."

Hope this helps. :]
-mk

(cc @stelmo)

### CODE REVIEW

File: stdlib/Test/src/Test.jl [hunk 1]
```diff
@@ -478,18 +484,24 @@ Test Broken
 
 julia> @test 2 + 2 == 4 skip=false
 Test Passed
+
+julia> @test 2 == 4 skip=false
+Test Passed
+
+julia>
```
Comment (stevengj) [inline stdlib/Test/src/Test.jl:491]: An example using the `expr` feature might be nice.  e.g. contrast
```jl
for i = 1:2; @test isodd(i); end
```
with
```jl
for i = 1:2; @test isodd(i) expr=:(isodd($i)); end
```

**Comment by stevengj** [comment]
Wouldn't it be more concise to write:
```jl
@eval @test looks_ok($x,$something)
```
?  That way you don't have to write the expression twice, and you get the same nice output.  For example:
```jl
julia> for i = 1:2; @test isodd(i); end
Test Failed at REPL[12]:1
  Expression: isodd(i)
ERROR: There was an error during testing

julia> for i = 1:2; @eval @test isodd($i); end    # more informative output
Test Failed at REPL[13]:1
  Expression: isodd(2)
ERROR: There was an error during testing
```
Since this works now, we don't need any new code, but a mention of this technique in the documentation could be useful.

File: stdlib/Test/src/Test.jl (hunk 1 above)
Comment (exaexa) [reply stdlib/Test/src/Test.jl:491]: Yap, will certainly add something.

**Comment by exaexa** [comment]
> Wouldn't it be more concise to write:
> 
> ```julia
> @eval @test looks_ok($x,$something)
> ```
> 

Ha, good point. I somehow failed to see this possibility. Is it perhaps possible to have this in a single macro? (I don't like wild `@eval`'s in the code.... but true that we eventually need at least a single `eval` for constructing the interpolated expression/string, so my opinion there are probably not an argument :D )

The other thing that I wanted to cover (and which is actually possible with this) is that sometimes the expression is too complex for other reasons, so I'd love to get stuff like:
```julia
@test totally_brutal_function_.(asd(asdasd(asdasdads(x)...)...)....)        simple_explanation="$x is suitable"
```
...which the `expr=` supports now, although not very semantically (it's not an expression). Would there perhaps be some other place in the structure (I guess in `struct Fail` where we could store the additional information?

> Since this works now, we don't need any new code, but a mention of this technique in the documentation could be useful.

I guess that's true. I'll try to convert this to docs. Any hint about the "additional nice information" as above would be welcome though.

**Comment by exaexa** [comment]
PS just adding the perf difference, the `@eval` solution unfortunately seems to compile a little bit.

```julia
julia> @time @testset "xxx" begin    # no annotation
         for i=1:1000
           @test i+1-1==i
         end
       end
Test Summary: | Pass  Total  Time
xxx           | 1000   1000  0.0s
  0.000496 seconds (9.20 k allocations: 431.352 KiB)
Test.DefaultTestSet("xxx", Any[], 1000, false, false, true, 1.683900836176925e9, 1.683900836177261e9, false, "REPL[8]")

julia> @time @testset "xxx" begin    # @eval solution
         for i=1:1000
           @eval @test $i+1-1==$i
         end
       end
Test Summary: | Pass  Total  Time
xxx           | 1000   1000  0.7s
  0.698131 seconds (339.09 k allocations: 15.991 MiB, 0.03% compilation time)
Test.DefaultTestSet("xxx", Any[], 1000, false, false, true, 1.683900842814456e9, 1.683900843512423e9, false, "REPL[9]")

julia> @time @testset "xxx" begin    # this PR
         for i=1:1000
           @test i+1-1==i expr=:($i+1-1==$i)
         end
       end
Test Summary: | Pass  Total  Time
xxx           | 1000   1000  0.1s
  0.050561 seconds (63.15 k allocations: 3.015 MiB)
Test.DefaultTestSet("xxx", Any[], 1000, false, false, true, 1.683900854823115e9, 1.683900854873504e9, false, "REPL[10]")
```

One other unwieldy thing is that `@eval` requires dollaring of all variables, which may and may not be fine in contexts.

**Comment by Seelengrab** [comment]
Could this be done by supporting  `$`-interpolation in `@test` itself, BenchmarkTools-style, ultimately recording the values themselves in the containing `TestSet` when ` record`ing the result?

**Comment by exaexa** [comment]
> Could this be done by supporting `$`-interpolation in `@test` itself, BenchmarkTools-style, ultimately recording the values themselves in the containing `TestSet` when `record`ing the result?

Yes, that would be best IMO, but with `@test` the problem is more difficult (there's lot of code transform happening to the expression on the way). If there's an easy way to just preprocess the expression into an $-interpolated form (for reporting) and a form without the $'s (for actual testing), that would be by far the best.

**Comment by exaexa** [comment]
Hi all, first, thanks everyone for time and input! In the end I settled with a custom `@xtest` macro that just supplies the extra "Expression" and "Evaluated" manually (which ain't super convenient for the programmer, but works and generates good generated-test reports). I'll eventually have a look into the possibility of interpolating the stuff with `$`'s (the main problem is actually _removing_ the `$`s so that they don't break test evaluation...). If anything nice surfaces, I'll reopen this.

**Pull request closed without merging**